### ISSUE TITLE
Question: Binary segmentation with weights

### ISSUE THREAD

**Issue by Nicholas-Schaub** [vlfeat/matconvnet#809]
This is not an issue but a question regarding binary classification. 

**Background**
I've read through the issues here and found that binary classification seems to work by setting labels to {-1,1}. I am trying to build something to do cell segmentation, and the cells are densely populated. When I train the network, the validation set converges pretty quickly to 0.15, which I was excited about this until I realized that about 15% of the pixels are labeled as cell edges. The network thought it was easier to just label the entire image as a cell, rather than segmenting individual cells.

**Question**
Based on what I've read and digging through the code, the {-1,1} can be thought of as weights for the logistic loss function since the logistic loss function seems to multiply the label with the prediction. If that's the case, I should be able to change the {-1,1} to something like {-1.5,8.5} and that should weight the labels roughly according to how frequently they occur. Is this correct?

I realize that this would mess up the binary error, and there might be other issues. I thought I would post and see what people who have used this more frequently think. I'm currently training a network using this, and I'll post results if there is interest.

**Comment by CarlosArteta** [comment]
If you want to explore balancing the training data, you can use the `instanceWeights` option of `vl_nnloss`. I think this would be the most straightforward approach. Alternatively, you can also balance the training data by balancing the batches; for example, your `getBatch` function could sample from each class in the proportions that you want.

**Comment by Nicholas-Schaub** [comment]
@CarlosArteta 
Instance weights seems to be static for all images, as far as I can tell. So for image segmentation, where every pixel in a segmented image may change for the next image, using instance weights doesn't seem to be an option. Can you set `instanceWeights` for every image?

The approach I suggested using above is working really well. I had to make a few modifications, and I changed `vl_nnloss` so that '`binaryerror` compared the sign of the class.

**Comment by CarlosArteta** [comment]
Sure, `instanceWeights` can have the same dimensions as your array of labels, so you can define a weight for each pixel in the segmentation. Like the labels, this can be unique for every image. I think this should give you the same good results you have now, but it will also be more general in case you then want to try a different loss, or change the number of classes, etc..

**Comment by Nicholas-Schaub** [comment]
Gotcha, I'll have to figure out how to do this. Thanks for the response.

**Comment by CarlosArteta** [comment]
No problem!
